File: desktop/ReadFlow/Core/SpeedCalculator.cs
using System;

namespace ReadFlow.Core
{
    /// <summary>
    /// Формули швидкості читання (специфікація, 4.7).
    ///
    /// Чисті функції без стану й без залежностей від WPF — те саме має бути
    /// буквально перенесене в Android-версію.
    /// </summary>
    public static class SpeedCalculator
    {
        private const int SecondsPerMinute = 60;

        /// <summary>Слів за хвилину.</summary>
        public static int WordsPerMinute(int wordsRead, decimal seconds)
        {
            return PerMinute(wordsRead, seconds);
        }

        /// <summary>
        /// Знаків за хвилину. Знаки — без пробілів: пробіли не вимовляються.
        /// </summary>
        public static int CharsPerMinute(int charsRead, decimal seconds)
        {
            return PerMinute(charsRead, seconds);
        }

        /// <summary>
        /// Скільки приблизно хвилин займе текст на заданій швидкості.
        ///
        /// Зворотна задача до <see cref="WordsPerMinute"/>: потрібна, щоб біля
        /// зразка написати «≈ 2 хв» для обраного класу. Округлення — «від нуля»
        /// на точному дробі, як і скрізь у розділі 4, і не менше однієї хвилини:
        /// «≈ 0 хв» не сказало б учителю нічого.
        /// </summary>
        public static int MinutesToRead(int words, int wordsPerMinute)
        {
            if (words <= 0 || wordsPerMinute <= 0)
            {
                return 0;
            }

            var minutes = (int)Math.Round(
                (decimal)words / wordsPerMinute, 0, MidpointRounding.AwayFromZero);

            return minutes < 1 ? 1 : minutes;
        }

        private static int PerMinute(int amount, decimal seconds)
        {
            if (seconds <= 0m || amount <= 0)
            {
                return 0;
            }

            // Ділення в decimal, а не в double: 45 слів за 120 с — це рівно 22.5,
            // справжня середина. На double результат залежав би від платформи
            // (та сама пастка, що й у середній довжині слова, див. 4.4).
            var perMinute = amount * (decimal)SecondsPerMinute / seconds;

            return (int)Math.Round(perMinute, 0, MidpointRounding.AwayFromZero);
        }
    }
}

File: android/app/src/main/java/net/readflow/core/SpeedCalculator.kt
package net.readflow.core

import java.math.BigDecimal
import java.math.RoundingMode

/**
 * Формули швидкості читання (`SPEC.md`, 4.7).
 *
 * Чисті функції без стану й без залежностей від Android — дзеркало
 * `desktop/ReadFlow/Core/SpeedCalculator.cs`. На однакових числах обидві
 * реалізації мусять давати однаковий результат.
 */
object SpeedCalculator {

    private val SECONDS_PER_MINUTE = BigDecimal(60)
    private val MILLIS_PER_SECOND = BigDecimal(1000)
    private val HUNDRED = BigDecimal(100)

    /** Слів за хвилину. */
    fun wordsPerMinute(wordsRead: Int, seconds: BigDecimal): Int = perMinute(wordsRead, seconds)

    /**
     * Знаків за хвилину. Знаки — **без пробілів**: пробіли не вимовляються,
     * тож у швидкості читання вголос їм місця немає.
     */
    fun charsPerMinute(charsRead: Int, seconds: BigDecimal): Int = perMinute(charsRead, seconds)

    /**
     * Відсоток помилок від прочитаного, до 0.1.
     *
     * Округлення — «від нуля» на точному дробі, як і все в розділі 4:
     * 5 помилок із 40 слів це рівно 12.5%, справжня середина, і на `Double`
     * результат залежав би від платформи.
     */
    fun errorPercent(errors: Int, wordsRead: Int): Double {
        if (errors <= 0 || wordsRead <= 0) {
            return 0.0
        }

        return BigDecimal(errors)
            .multiply(HUNDRED)
            .divide(BigDecimal(wordsRead), 1, RoundingMode.HALF_UP)
            .toDouble()
    }

    /**
     * Скільки приблизно хвилин займе текст на заданій швидкості.
     * Зворотна задача до [wordsPerMinute]; не менше однієї хвилини —
     * «≈ 0 хв» не сказало б учителю нічого.
     */
    fun minutesToRead(words: Int, wordsPerMinute: Int): Int {
        if (words <= 0 || wordsPerMinute <= 0) {
            return 0
        }

        val minutes = BigDecimal(words)
            .divide(BigDecimal(wordsPerMinute), 0, RoundingMode.HALF_UP)
            .toInt()

        return if (minutes < 1) 1 else minutes
    }

    /**
     * Мілісекунди в секунди **без втрати точності**.
     *
     * `millis / 1000.0` дав би `Double`, і тоді 22.5 у формулі швидкості
     * перестало б бути рівно 22.5 — рівно та пастка, від якої застерігає
     * `SPEC.md`, 4.4. Ділення на 1000 у `BigDecimal` завжди точне.
     */
    fun secondsOf(millis: Long): BigDecimal =
        BigDecimal(millis).divide(MILLIS_PER_SECOND)

    /** Секунди як точне ціле — для режиму A, де час береться з обраної тривалості. */
    fun secondsOf(seconds: Int): BigDecimal = BigDecimal(seconds)

    private fun perMinute(amount: Int, seconds: BigDecimal): Int {
        if (amount <= 0 || seconds.signum() <= 0) {
            return 0
        }

        // Ділення на точному дробі, а не на Double: 45 слів за 120 с — це рівно
        // 22.5, справжня середина, і на Double результат залежав би від платформи.
        return BigDecimal(amount)
            .multiply(SECONDS_PER_MINUTE)
            .divide(seconds, 0, RoundingMode.HALF_UP)
            .toInt()
    }
}
